include/dca/linalg/util/magma_queue.hpp

```diff
@@ -13,39 +13,66 @@
 #define DCA_LINALG_UTIL_MAGMA_QUEUE_HPP
 #ifdef DCA_HAVE_CUDA
 
+#include <cublas_v2.h>
 #include <cuda.h>
-#include <magma.h>
+#include <cusparse_v2.h>
+#include <magma_v2.h>
+
+#include "dca/linalg/util/cuda_stream.hpp"
 
 namespace dca {
 namespace linalg {
 namespace util {
 // dca::linalg::util::
 
-class MagmaQueue {
+class MagmaQueue : public linalg::util::CudaStream {
 public:
   MagmaQueue() {
-    magma_queue_create(&queue_);
+    cublasCreate(&cublas_handle_);
+    cusparseCreate(&cusparse_handle_);
+    int device;
+    cudaGetDevice(&device);
+    magma_queue_create_from_cuda(device, *this, cublas_handle_, cusparse_handle_, &queue_);
+  }
+
+  MagmaQueue(const MagmaQueue& rhs) = delete;
+  MagmaQueue& operator=(const MagmaQueue& rhs) = delete;
+
+  MagmaQueue(MagmaQueue&& rhs) : CudaStream(std::move(rhs)) {
+    swapMembers(rhs);
+  }
+
+  MagmaQueue& operator=(MagmaQueue&& rhs) {
```

Comment:
Again the semantics of operator= are pretty well established to not swap.  I'd expect the operator=(Whatever&& rhs) to leave an invalid moved from type.
You're going to need to justify this confusing operator= call to the base class.  This doesn't support using operator= in CudaStream since you aren't getting an operator overload anyway.

Also I see std::move(rhs) I would guess rhs is now invalid but then you call swapMembers.

Reply:
see comment above on move assignment.

Reply:
I don't see the problem in reusing the move constructor/assignment of the base class and having to define only the behavior of the new members.

Reply:
In other code bases this has led pretty directly to messes like this
class MySolver : public MPIObject, GPUObject, HDF5Object, OtherFunctionsWithState
Where MPIObject contains the communicator pointer and a few trivial member functions that would be more readable as my_mpi.rank() but now appear as rank().

GPUObject contains a stream pointer or cublas hande and some member functions....
Similar access patterns etc.

Reply:
I think it would be more consistent and simple to treat all these resources the same.  After all a MagmaQueue is also a cublas_handle + some stuff or a queue + some stuff.